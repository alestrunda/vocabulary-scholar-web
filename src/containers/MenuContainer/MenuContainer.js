/*global document*/

import React from 'react';
import PropTypes from 'prop-types';
import { menuItems } from '../../menuItems';
import Menu from '../../components/Menu';
import NavigationClose from 'material-ui/svg-icons/navigation/close';
import NavigationMenu from 'material-ui/svg-icons/navigation/menu';
import FloatingActionButton from 'material-ui/FloatingActionButton';
import classNames from 'classnames';

class MenuContainer extends React.Component {
  constructor(props) {
    super(props);
    this.buttonRef = React.createRef();
    this.handleClickOutside = this.handleClickOutside.bind(this);
  }

  state = {
    isVisible: false,
  };

  componentDidMount() {
    document.addEventListener('click', this.handleClickOutside);
  }

  componentWillUnmount() {
    document.removeEventListener('click', this.handleClickOutside);
  }

  handleClickOutside(e) {
    //close menu when clicked outside the menu button
    if (this.buttonRef.current.contains(e.target)) return;
    this.closeMenu();
  }

  closeMenu = () => {
    this.setState({
      isVisible: false,
    });
  };

  handleMenuButtonClick = () => {
    this.setState({
      isVisible: !this.state.isVisible, //toggle visibility
    });
  };

  render() {
    const menuLinks = menuItems.map(item => ({
      name: item.name,
      path: item.path,
      icon: item.icon,
    }));
    return (
      <aside
        className={classNames('page-navigation', this.props.className)}
        style={this.props.style}
      >
        <div
          ref={this.buttonRef}
          className={classNames('page-navigation__btn', {
            active: this.state.isVisible,
          })}
        >
          <FloatingActionButton
            style={{ width: 40, height: 40 }}
            iconStyle={{ width: 40, height: 40 }}
            backgroundColor="#fff"
            onClick={this.handleMenuButtonClick}
          >
            <NavigationMenu
              style={{
                width: 26,
                fill: '#00adb5',
                opacity: this.state.isVisible ? 0 : 1,
              }}
            />
            <NavigationClose
              className="navigation-icon-close"
              style={{
                fill: '#00adb5',
                width: 28,
                opacity: this.state.isVisible ? 1 : 0,
              }}
            />
          </FloatingActionButton>
        </div>
        <Menu
          className="page-navigation__menu"
          isVisible={this.state.isVisible}
          links={menuLinks}
          onClose={this.closeMenu}
        />
      </aside>
    );
  }
}

MenuContainer.propTypes = {
  className: PropTypes.string,
  style: PropTypes.object,
};

export default MenuContainer;
