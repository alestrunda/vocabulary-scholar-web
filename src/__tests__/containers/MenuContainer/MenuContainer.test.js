import React from 'react';
import { shallow } from 'enzyme';

import MenuContainer from '../../../containers/MenuContainer';

it('renders correctly', () => {
  const root = shallow(<MenuContainer />);
  expect(root).toMatchSnapshot();
});

it('opens on button click', () => {
  const root = shallow(<MenuContainer />);
  expect(root.state('isVisible')).toBe(false);
  root.find('FloatingActionButton').simulate('click');
  expect(root.state('isVisible')).toBe(true);
});
