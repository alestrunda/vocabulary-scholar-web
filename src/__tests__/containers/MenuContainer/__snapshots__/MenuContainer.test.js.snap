// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <MenuContainer />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <div
          className="page-navigation__btn"
        >
          <FloatingActionButton
            backgroundColor="#fff"
            disabled={false}
            iconStyle={
              Object {
                "height": 40,
                "width": 40,
              }
            }
            mini={false}
            onClick={[Function]}
            secondary={false}
            style={
              Object {
                "height": 40,
                "width": 40,
              }
            }
            zDepth={2}
          >
            <NavigationMenu
              style={
                Object {
                  "fill": "#00adb5",
                  "opacity": 1,
                  "width": 26,
                }
              }
            />
            <NavigationClose
              className="navigation-icon-close"
              style={
                Object {
                  "fill": "#00adb5",
                  "opacity": 0,
                  "width": 28,
                }
              }
            />
          </FloatingActionButton>
        </div>,
        <Menu
          className="page-navigation__menu"
          isVisible={false}
          links={
            Array [
              Object {
                "icon": <ActionHome
                  color="#fff"
                  style={
                    Object {
                      "display": "block",
                    }
                  }
                />,
                "name": "Home",
                "path": "/",
              },
              Object {
                "icon": <ActionBook
                  color="#fff"
                  style={
                    Object {
                      "display": "block",
                    }
                  }
                />,
                "name": "Lists",
                "path": "/list",
              },
              Object {
                "icon": <ActionAssignment
                  color="#fff"
                  style={
                    Object {
                      "display": "block",
                    }
                  }
                />,
                "name": "Lessons",
                "path": "/lesson",
              },
              Object {
                "icon": <EditorInsertChart
                  color="#fff"
                  style={
                    Object {
                      "display": "block",
                    }
                  }
                />,
                "name": "Overview",
                "path": "/overview",
              },
              Object {
                "icon": <ActionSettings
                  color="#fff"
                  style={
                    Object {
                      "display": "block",
                    }
                  }
                />,
                "name": "Settings",
                "path": "/settings",
              },
              Object {
                "icon": <ActionInfo
                  color="#fff"
                  style={
                    Object {
                      "display": "block",
                    }
                  }
                />,
                "name": "About",
                "path": "/about",
              },
            ]
          }
          onClose={[Function]}
        />,
      ],
      "className": "page-navigation",
      "style": undefined,
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": <FloatingActionButton
            backgroundColor="#fff"
            disabled={false}
            iconStyle={
              Object {
                "height": 40,
                "width": 40,
              }
            }
            mini={false}
            onClick={[Function]}
            secondary={false}
            style={
              Object {
                "height": 40,
                "width": 40,
              }
            }
            zDepth={2}
          >
            <NavigationMenu
              style={
                Object {
                  "fill": "#00adb5",
                  "opacity": 1,
                  "width": 26,
                }
              }
            />
            <NavigationClose
              className="navigation-icon-close"
              style={
                Object {
                  "fill": "#00adb5",
                  "opacity": 0,
                  "width": 28,
                }
              }
            />
          </FloatingActionButton>,
          "className": "page-navigation__btn",
        },
        "ref": Object {
          "current": null,
        },
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "backgroundColor": "#fff",
            "children": Array [
              <NavigationMenu
                style={
                  Object {
                    "fill": "#00adb5",
                    "opacity": 1,
                    "width": 26,
                  }
                }
              />,
              <NavigationClose
                className="navigation-icon-close"
                style={
                  Object {
                    "fill": "#00adb5",
                    "opacity": 0,
                    "width": 28,
                  }
                }
              />,
            ],
            "disabled": false,
            "iconStyle": Object {
              "height": 40,
              "width": 40,
            },
            "mini": false,
            "onClick": [Function],
            "secondary": false,
            "style": Object {
              "height": 40,
              "width": 40,
            },
            "zDepth": 2,
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "style": Object {
                  "fill": "#00adb5",
                  "opacity": 1,
                  "width": 26,
                },
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "className": "navigation-icon-close",
                "style": Object {
                  "fill": "#00adb5",
                  "opacity": 0,
                  "width": 28,
                },
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
          ],
          "type": [Function],
        },
        "type": "div",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "className": "page-navigation__menu",
          "isVisible": false,
          "links": Array [
            Object {
              "icon": <ActionHome
                color="#fff"
                style={
                  Object {
                    "display": "block",
                  }
                }
              />,
              "name": "Home",
              "path": "/",
            },
            Object {
              "icon": <ActionBook
                color="#fff"
                style={
                  Object {
                    "display": "block",
                  }
                }
              />,
              "name": "Lists",
              "path": "/list",
            },
            Object {
              "icon": <ActionAssignment
                color="#fff"
                style={
                  Object {
                    "display": "block",
                  }
                }
              />,
              "name": "Lessons",
              "path": "/lesson",
            },
            Object {
              "icon": <EditorInsertChart
                color="#fff"
                style={
                  Object {
                    "display": "block",
                  }
                }
              />,
              "name": "Overview",
              "path": "/overview",
            },
            Object {
              "icon": <ActionSettings
                color="#fff"
                style={
                  Object {
                    "display": "block",
                  }
                }
              />,
              "name": "Settings",
              "path": "/settings",
            },
            Object {
              "icon": <ActionInfo
                color="#fff"
                style={
                  Object {
                    "display": "block",
                  }
                }
              />,
              "name": "About",
              "path": "/about",
            },
          ],
          "onClose": [Function],
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
    ],
    "type": "aside",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <div
            className="page-navigation__btn"
          >
            <FloatingActionButton
              backgroundColor="#fff"
              disabled={false}
              iconStyle={
                Object {
                  "height": 40,
                  "width": 40,
                }
              }
              mini={false}
              onClick={[Function]}
              secondary={false}
              style={
                Object {
                  "height": 40,
                  "width": 40,
                }
              }
              zDepth={2}
            >
              <NavigationMenu
                style={
                  Object {
                    "fill": "#00adb5",
                    "opacity": 1,
                    "width": 26,
                  }
                }
              />
              <NavigationClose
                className="navigation-icon-close"
                style={
                  Object {
                    "fill": "#00adb5",
                    "opacity": 0,
                    "width": 28,
                  }
                }
              />
            </FloatingActionButton>
          </div>,
          <Menu
            className="page-navigation__menu"
            isVisible={false}
            links={
              Array [
                Object {
                  "icon": <ActionHome
                    color="#fff"
                    style={
                      Object {
                        "display": "block",
                      }
                    }
                  />,
                  "name": "Home",
                  "path": "/",
                },
                Object {
                  "icon": <ActionBook
                    color="#fff"
                    style={
                      Object {
                        "display": "block",
                      }
                    }
                  />,
                  "name": "Lists",
                  "path": "/list",
                },
                Object {
                  "icon": <ActionAssignment
                    color="#fff"
                    style={
                      Object {
                        "display": "block",
                      }
                    }
                  />,
                  "name": "Lessons",
                  "path": "/lesson",
                },
                Object {
                  "icon": <EditorInsertChart
                    color="#fff"
                    style={
                      Object {
                        "display": "block",
                      }
                    }
                  />,
                  "name": "Overview",
                  "path": "/overview",
                },
                Object {
                  "icon": <ActionSettings
                    color="#fff"
                    style={
                      Object {
                        "display": "block",
                      }
                    }
                  />,
                  "name": "Settings",
                  "path": "/settings",
                },
                Object {
                  "icon": <ActionInfo
                    color="#fff"
                    style={
                      Object {
                        "display": "block",
                      }
                    }
                  />,
                  "name": "About",
                  "path": "/about",
                },
              ]
            }
            onClose={[Function]}
          />,
        ],
        "className": "page-navigation",
        "style": undefined,
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": <FloatingActionButton
              backgroundColor="#fff"
              disabled={false}
              iconStyle={
                Object {
                  "height": 40,
                  "width": 40,
                }
              }
              mini={false}
              onClick={[Function]}
              secondary={false}
              style={
                Object {
                  "height": 40,
                  "width": 40,
                }
              }
              zDepth={2}
            >
              <NavigationMenu
                style={
                  Object {
                    "fill": "#00adb5",
                    "opacity": 1,
                    "width": 26,
                  }
                }
              />
              <NavigationClose
                className="navigation-icon-close"
                style={
                  Object {
                    "fill": "#00adb5",
                    "opacity": 0,
                    "width": 28,
                  }
                }
              />
            </FloatingActionButton>,
            "className": "page-navigation__btn",
          },
          "ref": Object {
            "current": null,
          },
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "backgroundColor": "#fff",
              "children": Array [
                <NavigationMenu
                  style={
                    Object {
                      "fill": "#00adb5",
                      "opacity": 1,
                      "width": 26,
                    }
                  }
                />,
                <NavigationClose
                  className="navigation-icon-close"
                  style={
                    Object {
                      "fill": "#00adb5",
                      "opacity": 0,
                      "width": 28,
                    }
                  }
                />,
              ],
              "disabled": false,
              "iconStyle": Object {
                "height": 40,
                "width": 40,
              },
              "mini": false,
              "onClick": [Function],
              "secondary": false,
              "style": Object {
                "height": 40,
                "width": 40,
              },
              "zDepth": 2,
            },
            "ref": null,
            "rendered": Array [
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "class",
                "props": Object {
                  "style": Object {
                    "fill": "#00adb5",
                    "opacity": 1,
                    "width": 26,
                  },
                },
                "ref": null,
                "rendered": null,
                "type": [Function],
              },
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "class",
                "props": Object {
                  "className": "navigation-icon-close",
                  "style": Object {
                    "fill": "#00adb5",
                    "opacity": 0,
                    "width": 28,
                  },
                },
                "ref": null,
                "rendered": null,
                "type": [Function],
              },
            ],
            "type": [Function],
          },
          "type": "div",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "className": "page-navigation__menu",
            "isVisible": false,
            "links": Array [
              Object {
                "icon": <ActionHome
                  color="#fff"
                  style={
                    Object {
                      "display": "block",
                    }
                  }
                />,
                "name": "Home",
                "path": "/",
              },
              Object {
                "icon": <ActionBook
                  color="#fff"
                  style={
                    Object {
                      "display": "block",
                    }
                  }
                />,
                "name": "Lists",
                "path": "/list",
              },
              Object {
                "icon": <ActionAssignment
                  color="#fff"
                  style={
                    Object {
                      "display": "block",
                    }
                  }
                />,
                "name": "Lessons",
                "path": "/lesson",
              },
              Object {
                "icon": <EditorInsertChart
                  color="#fff"
                  style={
                    Object {
                      "display": "block",
                    }
                  }
                />,
                "name": "Overview",
                "path": "/overview",
              },
              Object {
                "icon": <ActionSettings
                  color="#fff"
                  style={
                    Object {
                      "display": "block",
                    }
                  }
                />,
                "name": "Settings",
                "path": "/settings",
              },
              Object {
                "icon": <ActionInfo
                  color="#fff"
                  style={
                    Object {
                      "display": "block",
                    }
                  }
                />,
                "name": "About",
                "path": "/about",
              },
            ],
            "onClose": [Function],
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
      ],
      "type": "aside",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromProps": Object {
            "hasShouldComponentUpdateBug": false,
          },
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
  Symbol(enzyme.__childContext__): null,
}
`;
